File: python_orig/make_image.py
import numpy as np
def create_complex_grid(xlim, ylim, resolution):
    """
    Create a 2D numpy array of complex numbers.

    Parameters:
    xlim (tuple): The limits for the x-axis (real part), e.g., (-1, 1).
    ylim (tuple): The limits for the y-axis (imaginary part), e.g., (-1, 1).
    resolution (int): The number of points along each axis.

    Returns:
    numpy.ndarray: A 2D array of complex numbers.
    """
    # Create linearly spaced arrays for the real and imaginary parts
    real = np.linspace(xlim[0], xlim[1], resolution)
    imag = np.linspace(ylim[0], ylim[1], resolution)

    # Create a meshgrid, which forms a coordinate matrix from the real and imaginary parts
    real_grid, imag_grid = np.meshgrid(real, imag)

    # Combine the real and imaginary parts to form complex numbers
    complex_grid = real_grid + 1j * imag_grid

    return complex_grid

def make_image(xlim: tuple, ylim: tuple, res: int, func):

    inputs = create_complex_grid(xlim,ylim,res)
    outputs = np.zeros(inputs.shape, dtype=complex)

    #generate outputs
    for i,j in np.ndindex(outputs.shape):
        outputs[i,j] = func(inputs[i,j])

    return outputs

File: legacy/make_image.cpp
#include <iostream>
#include <vector>
#include <complex>

// Function to create a linearly spaced vector
std::vector<double> linspace(double start, double end, int num) {
    std::vector<double> linspaced;
    double delta = (end - start) / (num - 1);

    for (int i = 0; i < num - 1; ++i) {
        linspaced.push_back(start + delta * i);
    }
    linspaced.push_back(end); // Ensure that end is included

    return linspaced;
}

// Function to create a 2D grid of complex numbers
std::vector<std::vector<std::complex<double>>> create_complex_grid(std::pair<double,double> xlim, std::pair<double,double> ylim, int resolution) {
    std::vector<double> real = linspace(xlim.first, xlim.second, resolution);
    std::vector<double> imag = linspace(ylim.first, ylim.second, resolution);

    std::vector<std::vector<std::complex<double>>> complex_grid(resolution, std::vector<std::complex<double>>(resolution));

    for (int i = 0; i < resolution; ++i) {
        for (int j = 0; j < resolution; ++j) {
            complex_grid[i][j] = std::complex<double>(real[i], imag[j]);
        }
    }

    return complex_grid;
}

// Function to apply a complex function to each element of the grid
std::vector<std::vector<std::complex<double>>> make_image(std::pair<double,double> xlim, std::pair<double,double> ylim, int res, std::complex<double> (*func)(const std::complex<double>&)) {
    auto inputs = create_complex_grid(xlim, ylim, res);
    std::vector<std::vector<std::complex<double>>> outputs(res, std::vector<std::complex<double>>(res));

    for (int i = 0; i < res; ++i) {
        for (int j = 0; j < res; ++j) {
            outputs[i][j] = func(inputs[i][j]);
        }
    }

    return outputs;
}
